tempodb/encoding/vparquet2/schema.go

```diff
@@ -0,0 +1,636 @@
+	FieldSpanAttrValInt    = "rs.list.element.ss.list.element.Spans.list.element.Attrs.list.element.ValueInt"
+	FieldSpanAttrValDouble = "rs.list.element.ss.list.element.Spans.list.element.Attrs.list.element.ValueDouble"
+	FieldSpanAttrValBool   = "rs.list.element.ss.list.element.Spans.list.element.Attrs.list.element.ValueBool"
+)
+
+var (
+	jsonMarshaler = new(jsonpb.Marshaler)
+
+	labelMappings = map[string]string{
+		LabelRootSpanName:     "RootSpanName",
+		LabelRootServiceName:  "RootServiceName",
+		LabelServiceName:      "rs.list.element.Resource.ServiceName",
+		LabelCluster:          "rs.list.element.Resource.Cluster",
+		LabelNamespace:        "rs.list.element.Resource.Namespace",
+		LabelPod:              "rs.list.element.Resource.Pod",
+		LabelContainer:        "rs.list.element.Resource.Container",
+		LabelK8sClusterName:   "rs.list.element.Resource.K8sClusterName",
+		LabelK8sNamespaceName: "rs.list.element.Resource.K8sNamespaceName",
+		LabelK8sPodName:       "rs.list.element.Resource.K8sPodName",
+		LabelK8sContainerName: "rs.list.element.Resource.K8sContainerName",
+		LabelName:             "rs.list.element.ss.list.element.Spans.list.element.Name",
+		LabelHTTPMethod:       "rs.list.element.ss.list.element.Spans.list.element.HttpMethod",
+		LabelHTTPUrl:          "rs.list.element.ss.list.element.Spans.list.element.HttpUrl",
+		LabelHTTPStatusCode:   "rs.list.element.ss.list.element.Spans.list.element.HttpStatusCode",
+		LabelStatusCode:       "rs.list.element.ss.list.element.Spans.list.element.StatusCode",
+	}
+)
+
+type Attribute struct {
+	Key string `parquet:",snappy,dict"`
+
+	// This is a bad design that leads to millions of null values. How can we fix this?
+	Value       *string  `parquet:",dict,snappy,optional"`
+	ValueInt    *int64   `parquet:",snappy,optional"`
+	ValueDouble *float64 `parquet:",snappy,optional"`
+	ValueBool   *bool    `parquet:",snappy,optional"`
+	ValueKVList string   `parquet:",snappy,optional"`
+	ValueArray  string   `parquet:",snappy,optional"`
+}
+
+type EventAttribute struct {
+	Key   string `parquet:",snappy,dict"`
+	Value []byte `parquet:",snappy"` // Was json-encoded data, is now proto encoded data
+}
+
+type Event struct {
+	TimeUnixNano           uint64           `parquet:",delta"`
+	Name                   string           `parquet:",snappy"`
+	Attrs                  []EventAttribute `parquet:",list"`
+	DroppedAttributesCount int32            `parquet:",snappy,delta"`
+	Test                   string           `parquet:",snappy,dict,optional"` // Always empty for testing
+}
+
+// nolint:revive
+// Ignore field naming warnings
+type Span struct {
+	// SpanID is []byte to save space. It doesn't need to be user
+	// friendly like trace ID, and []byte is half the size of string.
+	SpanID                 []byte      `parquet:","`
+	ParentSpanID           []byte      `parquet:","`
+	ParentID               int32       `parquet:",delta"`
```

Comment:
any thoughts on the delta compression chosen here? since we're not populating this with data it's quite hard to make guesses about what would be most performant, but I thought maybe you had some ideas to share.

Reply:
My reasoning behind using delta here is the following:
* The numerical span IDs and parent IDs will be derived from the row of the span. The biggest difference between the parent ID of two spans is roughly the number of spans in the trace
* Another assumption is that closely related spans are also close together in the trace data structure, which also means that the difference in their parent IDs are also expected to be small. I’m not sure if this is true though

Reply:
yup, i had similar considerations.